src/Blockcore.Indexer.Core/Storage/Mongo/BlockRewindOperation.cs

```diff
@@ -61,18 +65,44 @@ public static async Task RewindBlockAsync(this MongoData storage, long blockInde
       Task<DeleteResult> addressHistoryComputed =
          storage.AddressHistoryComputedTable.DeleteManyAsync(addrCompHistFilter);
 
+      await Task.WhenAll( output, transactions, addressComputed, addressHistoryComputed);
+
+      await MergeRewindInputsToUnspentTransactionsAsync(storage, blockIndex);
+
+      FilterDefinition<InputTable> inputFilter =
+         Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, blockIndex);
+
+      var inputs = storage.InputTable.DeleteManyAsync(inputFilter);
+
       FilterDefinition<UnspentOutputTable> unspentOutputFilter =
          Builders<UnspentOutputTable>.Filter.Eq(utxo => utxo.BlockIndex, blockIndex);
       Task<DeleteResult> unspentOutput = storage.UnspentOutputTable.DeleteManyAsync(unspentOutputFilter);
 
-      await Task.WhenAll(unspentOutput, output, transactions, addressComputed, addressHistoryComputed);
+      await Task.WhenAll( inputs, unspentOutput);
+   }
 
-      await MergeRewindInputsToUnspentTransactionsAsync(storage, blockIndex);
+   static Task StoreRewindBlockAsync(MongoDb storage, uint blockIndex)
+   {
+      var blockTask = storage.BlockTable.FindAsync(_ => _.BlockIndex == blockIndex);
+      var inputsTask = storage.InputTable.FindAsync(_ => _.BlockIndex == blockIndex);
+      var outputsTask = storage.OutputTable.FindAsync(_ => _.BlockIndex == blockIndex);
+      var transactionIdsTask = storage.TransactionBlockTable.FindAsync(_ => _.BlockIndex == blockIndex);
 
-      FilterDefinition<InputTable> inputFilter =
-         Builders<InputTable>.Filter.Eq(addr => addr.BlockIndex, blockIndex);
+      Task.WhenAll(blockTask, inputsTask, outputsTask, transactionIdsTask);
 
-      await storage.InputTable.DeleteManyAsync(inputFilter);
+      var block = blockTask.Result.Single();
+
+      var reorgBlock = new ReorgBlockTable
```

Comment:
We should probably add another property here for the time the reorg happened.
Its a valid scenario that the same block will be reorged and go back to main chain and reorg again etc...
A network under attack with high malicious % of stake may experience that behavior